t/test-rdata_to_str.c

```diff
@@ -291,42 +291,129 @@ struct test tdata[] = {
+	/* draft-ietf-dnsop-svcb-https-08 */
+	{ /* appendix D, figure 1 */
+		.input = "\x00\x00"				/* priority */
+		    "\x03""foo\x07""example\x03""com\x00",	/* target */
+		.input_len = 19,
+		.rrclass = WDNS_CLASS_IN,
+		.rrtype = WDNS_TYPE_HTTPS,
+		.expected = "0 foo.example.com.",
+	},
+
+	{ /* appendix D, figure 2 */
+		.input = "\x00\x01"				/* priority */
+		    "\x00",					/* target */
+		.input_len = 3,
+		.rrclass = WDNS_CLASS_IN,
+		.rrtype = WDNS_TYPE_HTTPS,
+		.expected = "1 .",
+	},
+
+	{ /* appendix D, figure 3 */
+		.input = "\x00\x10"				/* priority */
+		    "\x03""foo\x07""example\x03""com\x00"	/* target */
+		    "\x00\x03"					/* port */
+		    "\x00\x02"					/* length.. */
+		    "\x00\x35",					/* ..value */
+		.input_len = 25,
+		.rrclass = WDNS_CLASS_IN,
+		.rrtype = WDNS_TYPE_SVCB,
+		.expected = "16 foo.example.com. port=53",
+	},
+
+	{ /* appendix D, figure 4 */
+		.input = "\x00\x01"				/* priority */
+		    "\x03""foo\x07""example\x03""com\x00"	/* target */
+		    "\x02\x9b"					/* 667 */
+		    "\x00\x05"					/* length.. */
+		    "hello",					/* ..value */
+		.input_len = 28,
+		.rrclass = WDNS_CLASS_IN,
+		.rrtype = WDNS_TYPE_SVCB,
+		.expected = "1 foo.example.com. key667=\"hello\"",
+	},
+
+	/* appendix D, figure 5 is omitted as it fails the round trip test */
+
+	{ /* appendix D, figure 6 */
+		.input = "\x00\x01"				/* priority */
+		    "\x03""foo\x07""example\x03""com\x00"	/* target */
+		    "\x00\x06"					/* ipv6hint */
+		    "\x00\x20"					/* length.. */
+		    "\x20\x01\x0d\xb8\x00\x00\x00\x00"		/* ..value */
+		    "\x00\x00\x00\x00\x00\x00\x00\x01"
+		    "\x20\x01\x0d\xb8\x00\x00\x00\x00"
+		    "\x00\x00\x00\x00\x00\x53\x00\x01",
+		.input_len = 55,
+		.rrclass = WDNS_CLASS_IN,
+		.rrtype = WDNS_TYPE_SVCB,
+		.expected = "1 foo.example.com. ipv6hint=2001:db8::1,2001:db8::53:1",
+	},
+
+	{ /* appendix D, figure 7 */
+		.input = "\x00\x01"				/* priority */
+		    "\x07""example\x03""com\x00"		/* target */
+		    "\x00\x06"					/* ipv6hint */
+		    "\x00\x10"					/* length.. */
+		    "\x00\x00\x00\x00\x00\x00\x00\x00"		/* ..value */
+		    "\x00\x00\xff\xff\xc6\x33\x64\x64",
+		.input_len = 35,
+		.rrclass = WDNS_CLASS_IN,
+		.rrtype = WDNS_TYPE_SVCB,
+		.expected = "1 example.com. ipv6hint=::ffff:198.51.100.100",
+	},
+
+	/* appendix D, figure 8 is omitted as it relies on re-ordering params */
+
+	{ /* appendix D, figure 9 (reverse test) */
+                .input = "\x00\x10"				/* priority */
+                    "\x03""foo\x07""example\x03org\x00"		/* target */
+                    "\x00\x01"					/* alpn */
+                    "\x00\x0c"					/* length.. */
+                    "\x08"					/* ..value */
+                    "f\\oo,bar"
+                    "\x02"
+                    "h2",
+                .input_len = 35,
+                .rrclass = WDNS_CLASS_IN,
+                .rrtype = WDNS_TYPE_SVCB,
+                .expected = "16 foo.example.org. alpn=\"f\\\\oo\\,bar,h2\"",
```

Comment:
This needs to match:
```
		.input = "16 foo.example.org. alpn=\"f\\\\\\\\oo\\\\,bar,h2\"",
```
in the corresponding test case in `t/test-str_to_rdata.c`


In my estimation, we're nearing the point where C preprocessor hackery along the lines of:
```
#define QUOTE(...) #__VA_ARGS__
...
   .expected = QUOTE(16 foo.example.org. alpn="f\\\oo\\,bar,h2"),

```
is more ergonomic than manually stacking extra layers of escaping into C string literals. Note that this allows the argument to the `QUOTE` macro to exactly match the visible text from the standard, which may make this a net win.

(I've used the above hack to good effect in quoting JSON blobs as C literals. It works quite well as long as the quoted data does not include a `)` character)